kedro/framework/context/context.py

```diff
@@ -159,65 +159,47 @@ def _update_nested_dict(old_dict: Dict[Any, Any], new_dict: Dict[Any, Any]) -> N
-            hook_manager: The ``PluginManager`` to activate hooks, supplied by the session.
-            env: Optional argument for configuration default environment to be used
-                for running the pipeline. If not specified, it defaults to "local".
-            extra_params: Optional dictionary containing extra project parameters.
-                If specified, will update (and therefore take precedence over)
-                the parameters retrieved from the project configuration.
-        """
-        self._project_path = Path(project_path).expanduser().resolve()
-        self._package_name = package_name
-        self._config_loader = config_loader
-        self._env = env
-        self._extra_params = deepcopy(extra_params)
-        self._hook_manager = hook_manager
-
-    @property  # type: ignore
-    def env(self) -> Optional[str]:
-        """Property for the current Kedro environment.
-
-        Returns:
-            Name of the current Kedro environment.
+    package_name: InitVar[str]
+    project_path: InitVar[Union[Path, str]]
+    config_loader: ConfigLoader
+    hook_manager: InitVar[PluginManager]
+    env: Optional[str] = None
+    extra_params: InitVar[Dict[str, Any]] = None
 
-        """
-        return self._env
+    """Create a context object by providing the root of a Kedro project and
+    the environment configuration subfolders
+    (see ``kedro.config.ConfigLoader``)
 
-    @property
-    def project_path(self) -> Path:
-        """Read-only property containing Kedro's root project directory.
+    Raises:
+        KedroContextError: If there is a mismatch
+            between Kedro project version and package version.
 
-        Returns:
-            Project directory.
+    Args:
+        package_name: Package name for the Kedro project the context is
+            created for.
+        project_path: Project path to define the context for.
+        config_loader: Kedro's ``ConfigLoader`` for loading the configuration files.
+        hook_manager: The ``PluginManager`` to activate hooks, supplied by the session.
+        env: Optional argument for configuration default environment to be used
+            for running the pipeline. If not specified, it defaults to "local".
+        extra_params: Optional dictionary containing extra project parameters.
+            If specified, will update (and therefore take precedence over)
+            the parameters retrieved from the project configuration.
+    """
 
-        """
-        return self._project_path
+    def __post_init__(self, package_name, project_path, hook_manager, extra_params):
+        object.__setattr__(self, "_package_name", package_name)
+        object.__setattr__(
+            self, "project_path", Path(project_path).expanduser().resolve()
+        )
+        object.__setattr__(self, "_hook_manager", hook_manager)
+        object.__setattr__(self, "_extra_params", deepcopy(extra_params))
```

Comment:
What's the logic behind which attributes are added here and the ones that aren't? For example, why is `package_name` added but not `env`?

Reply:
Good questions!

Personally, I am not too worried about it, but it is still a change in API since we now open up the possibility to overwrite it (though we can argue users could overwrite the `self._xxxx` if they are hacky.

Reply:
The current implementation is approach b, which works functionally, but it will also break IDE support & linters they seem to have trouble understand `obj.__setattr__(self, 'a', 1)` has the same semantic as `self.a = 1`. Things like code navigation no longer works.

My hesitation is seems we didn't get many benefits with `dataclass` and do we really need it.

Reply:
Would love to hear some opinion from @idanov